### PR TITLE
support TP-only parallelism

### PR DESCRIPTION
**Pull request by tianyu-l**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137

With simply fixes, we can use TP/SP alone without FSDP:
1. enable gradient scaling only when FSDP is enabled
2. use `torch.nn.utils.clip_grad_norm_` for gradient clipping when FSDP is disabled

Screenshot 2024-03-13 at 3 12 49 PM

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -240,7 +240,12 @@ def main(job_config: JobConfig):
 
             # clip gradients (after unscaling gradients of the optimizer's params)
             scaler.unscale_(optimizer)
-            model.clip_grad_norm_(job_config.training.max_norm)
+            if isinstance(model, FSDP):
+                model.clip_grad_norm_(job_config.training.max_norm)
```
Comment (wanchaol) [inline train.py:244]: I think this can be simplified once FSDP2 kicked in. cc @awgu
